### ISSUE TITLE
Virtiofs performance issue with huge files

### ISSUE THREAD

**Issue by ganeshmaharaj** [kata-containers/runtime#2815]
# Description of problem
virtiofs seems to have a huge performance impact against 9p when testing fio with huge files.

# Expected result
There is no performance impact when testing with huge files.

# Actual result
Logs from the various scenarios running fio with the below config
`fio --randrepeat=1 --ioengine=libaio --direct=1 --gtod_reduce=1 --name=test --filename=random_read_write.fio --bs=4k --iodepth=64 --size=4G --readwrite=randrw --rwmixread=75`

FIO TEST with 9p
```
[root@fiotest-brhcb /]# date; fio --randrepeat=1 --ioengine=libaio --direct=1 --gtod_reduce=1 --name=test --filename=random_read_write.fio --bs=4k --iodepth=64 --size=4G --readwrite=randrw --rwmixread=75
Thu Jul  2 20:38:00 UTC 2020
test: (g=0): rw=randrw, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=64
fio-3.19
Starting 1 process
test: Laying out IO file (1 file / 4096MiB)
Jobs: 1 (f=1): [m(1)][100.0%][r=19.0MiB/s,w=6573KiB/s][r=4872,w=1643 IOPS][eta 00m:00s]
test: (groupid=0, jobs=1): err= 0: pid=132: Thu Jul  2 20:41:47 2020
  read: IOPS=6492, BW=25.4MiB/s (26.6MB/s)(3070MiB/121047msec)
   bw (  KiB/s): min= 7712, max=37916, per=100.00%, avg=26013.00, stdev=5136.55, samples=241
   iops        : min= 1928, max= 9479, avg=6503.05, stdev=1284.22, samples=241
  write: IOPS=2169, BW=8679KiB/s (8888kB/s)(1026MiB/121047msec); 0 zone resets
   bw (  KiB/s): min= 3025, max=12590, per=100.00%, avg=8693.31, stdev=1725.98, samples=241
   iops        : min=  756, max= 3147, avg=2173.12, stdev=431.55, samples=241
  cpu          : usr=3.61%, sys=45.31%, ctx=1025253, majf=0, minf=6
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=785920,262656,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=25.4MiB/s (26.6MB/s), 25.4MiB/s-25.4MiB/s (26.6MB/s-26.6MB/s), io=3070MiB (3219MB), run=121047-121047msec
  WRITE: bw=8679KiB/s (8888kB/s), 8679KiB/s-8679KiB/s (8888kB/s-8888kB/s), io=1026MiB (1076MB), run=121047-121047msec
```
FIO TEST with virtiofs cache=always virtio_fs_cache_size=1024
```
[root@fiotest-r28wj /]# date; fio --randrepeat=1 --ioengine=libaio --direct=1 --gtod_reduce=1 --name=test --filename=random_read_write.fio --bs=4k --iodepth=64 --size=4G --readwrite=randrw --rwmixread=75
Thu Jul  2 20:50:33 UTC 2020
test: (g=0): rw=randrw, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=64
fio-3.19
Starting 1 process
test: Laying out IO file (1 file / 4096MiB)
Jobs: 1 (f=1): [m(1)][99.6%][r=8439KiB/s,w=2883KiB/s][r=2109,w=720 IOPS][eta 00m:09s]
test: (groupid=0, jobs=1): err= 0: pid=37: Thu Jul  2 21:27:10 2020
  read: IOPS=380, BW=1523KiB/s (1559kB/s)(3070MiB/2064461msec)
   bw (  KiB/s): min=  817, max=16735, per=99.96%, avg=1521.38, stdev=456.83, samples=4086
   iops        : min=  204, max= 4183, avg=380.02, stdev=114.23, samples=4086
  write: IOPS=127, BW=509KiB/s (521kB/s)(1026MiB/2064461msec); 0 zone resets
   bw (  KiB/s): min=  220, max= 5541, per=100.00%, avg=508.13, stdev=161.71, samples=4086
   iops        : min=   55, max= 1385, avg=126.61, stdev=40.46, samples=4086
  cpu          : usr=1.36%, sys=25.10%, ctx=1615840, majf=0, minf=42922
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=785920,262656,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=1523KiB/s (1559kB/s), 1523KiB/s-1523KiB/s (1559kB/s-1559kB/s), io=3070MiB (3219MB), run=2064461-2064461msec
  WRITE: bw=509KiB/s (521kB/s), 509KiB/s-509KiB/s (521kB/s-521kB/s), io=1026MiB (1076MB), run=2064461-2064461msec
```

FIO TEST with virtiofs cache=always virtio_fs_cache_size=0
```
[root@fiotest-4vln9 /]# date; fio --randrepeat=1 --ioengine=libaio --direct=1 --gtod_reduce=1 --name=test --filename=random_read_write.fio --bs=4k --iodepth=64 --size=4G --readwrite=randrw --rwmixread=75
Thu Jul  2 21:36:54 UTC 2020
test: (g=0): rw=randrw, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=64
fio-3.19
Starting 1 process
test: Laying out IO file (1 file / 4096MiB)


Jobs: 1 (f=1): [m(1)][100.0%][r=8678KiB/s,w=2834KiB/s][r=2169,w=708 IOPS][eta 00m:00s]
test: (groupid=0, jobs=1): err= 0: pid=35: Thu Jul  2 22:02:00 2020
  read: IOPS=1556, BW=6226KiB/s (6376kB/s)(3070MiB/504913msec)
   bw (  KiB/s): min=   23, max=16550, per=100.00%, avg=6920.50, stdev=2886.99, samples=906
   iops        : min=    5, max= 4137, avg=1729.76, stdev=721.75, samples=906
  write: IOPS=520, BW=2081KiB/s (2131kB/s)(1026MiB/504913msec); 0 zone resets
   bw (  KiB/s): min=   39, max= 5389, per=100.00%, avg=2315.17, stdev=971.38, samples=905
   iops        : min=    9, max= 1347, avg=578.45, stdev=242.87, samples=905
  cpu          : usr=2.05%, sys=24.84%, ctx=545866, majf=0, minf=7
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=785920,262656,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=6226KiB/s (6376kB/s), 6226KiB/s-6226KiB/s (6376kB/s-6376kB/s), io=3070MiB (3219MB), run=504913-504913msec
  WRITE: bw=2081KiB/s (2131kB/s), 2081KiB/s-2081KiB/s (2131kB/s-2131kB/s), io=1026MiB (1076MB), run=504913-504913msec
```

FIO TEST with virtiofs cache=always virtio_fs_cache_size=0
```
[root@fiotest-pblvb /]# date; fio --randrepeat=1 --ioengine=libaio --direct=1 --gtod_reduce=1 --name=test --filename=random_read_write.fio --bs=4k --iodepth=64 --size=4G --readwrite=randrw --rwmixread=75
Thu Jul  2 22:55:47 UTC 2020
test: (g=0): rw=randrw, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=64
fio-3.19
Starting 1 process
test: Laying out IO file (1 file / 4096MiB)
Jobs: 1 (f=1): [m(1)][99.7%][r=30.3MiB/s,w=10.1MiB/s][r=7749,w=2578 IOPS][eta 00m:01s]
test: (groupid=0, jobs=1): err= 0: pid=36: Thu Jul  2 23:12:54 2020
  read: IOPS=2119, BW=8477KiB/s (8680kB/s)(3070MiB/370851msec)
   bw (  KiB/s): min=    7, max=43144, per=100.00%, avg=24752.51, stdev=11906.09, samples=252
   iops        : min=    1, max=10786, avg=6187.75, stdev=2976.60, samples=252
  write: IOPS=708, BW=2833KiB/s (2901kB/s)(1026MiB/370851msec); 0 zone resets
   bw (  KiB/s): min=   23, max=14483, per=100.00%, avg=8304.96, stdev=3958.14, samples=251
   iops        : min=    5, max= 3620, avg=2075.91, stdev=989.61, samples=251
  cpu          : usr=1.20%, sys=11.82%, ctx=214808, majf=0, minf=8
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=785920,262656,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
   READ: bw=8477KiB/s (8680kB/s), 8477KiB/s-8477KiB/s (8680kB/s-8680kB/s), io=3070MiB (3219MB), run=370851-370851msec
  WRITE: bw=2833KiB/s (2901kB/s), 2833KiB/s-2833KiB/s (2901kB/s-2901kB/s), io=1026MiB (1076MB), run=370851-370851msec

```

**Comment by ganeshmaharaj** [comment]
@jcvenegas @amshinde

**Comment by rhvgoyal** [comment]
@ganeshmaharaj I have tried this in a VM and my observations are as follows.

- I am not seeing the slowdown you are seeing. In fact cache=none is fastest with virtiofs. cache=always slows it down surprisingly. I am not sure why. Will have to debug. Even dax seems to slow it down. Part of the reason is that it is random IO and part of it is that this is old dax code base and some of the locking improvements might not be here. 

Anyway, I tried following configuraiton with fio command you sent and following are my observations.

9p (cache=none)
-----------------------

> Run status group 0 (all jobs):
   READ: bw=43.1MiB/s (45.2MB/s), 43.1MiB/s-43.1MiB/s (45.2MB/s-45.2MB/s), io=
3070MiB (3219MB), run=71260-71260msec
  WRITE: bw=14.4MiB/s (15.1MB/s), 14.4MiB/s-14.4MiB/s (15.1MB/s-15.1MB/s), io=
1026MiB (1076MB), run=71260-71260msec

9p (cache=loose)
-----------------------

>    READ: bw=43.1MiB/s (45.2MB/s), 43.1MiB/s-43.1MiB/s (45.2MB/s-45.2MB/s), io=
3070MiB (3219MB), run=71182-71182msec
  WRITE: bw=14.4MiB/s (15.1MB/s), 14.4MiB/s-14.4MiB/s (15.1MB/s-15.1MB/s), io=
1026MiB (1076MB), run=71182-71182msec

virtiofs (cache=none)
----------------------------

> Run status group 0 (all jobs):
   READ: bw=211MiB/s (222MB/s), 211MiB/s-211MiB/s (222MB/s-222MB/s), io=3070Mi
B (3219MB), run=14530-14530msec
  WRITE: bw=70.6MiB/s (74.0MB/s), 70.6MiB/s-70.6MiB/s (74.0MB/s-74.0MB/s), io=
1026MiB (1076MB), run=14530-14530msec

virtiofs(cache=none, dax)
---------------------------------

>    READ: bw=105MiB/s (110MB/s), 105MiB/s-105MiB/s (110MB/s-110MB/s), io=3070Mi
B (3219MB), run=29219-29219msec
  WRITE: bw=35.1MiB/s (36.8MB/s), 35.1MiB/s-35.1MiB/s (36.8MB/s-36.8MB/s), io=
1026MiB (1076MB), run=29219-29219msec

virtiofs(cache=always)
------------------------------

> Run status group 0 (all jobs):
   READ: bw=44.9MiB/s (47.0MB/s), 44.9MiB/s-44.9MiB/s (47.0MB/s-47.0MB/s), io=
3070MiB (3219MB), run=68423-68423msec
  WRITE: bw=14.0MiB/s (15.7MB/s), 14.0MiB/s-14.0MiB/s (15.7MB/s-15.7MB/s), io=
1026MiB (1076MB), run=68423-68423msec

virtiofs(cache=always, dax)
------------------------------------

>    READ: bw=99.0MiB/s (104MB/s), 99.0MiB/s-99.0MiB/s (104MB/s-104MB/s), io=307
0MiB (3219MB), run=30998-30998msec
  WRITE: bw=33.1MiB/s (34.7MB/s), 33.1MiB/s-33.1MiB/s (34.7MB/s-34.7MB/s), io=
1026MiB (1076MB), run=30998-30998msec


In summary, cache=none is very fast for this workload. dax is in the middle and cache=always seems very slow and is almost same as 9p. 

I will spend some time figuring out why cache=always is slow. Otherwise in rest of the cases we are much faster than 9p (atleast in my tests).

I am using 5.8-rc4 kernel from virtio-fs-dev branch from gitlab and 5.0 qemu from virtio-fs-dev branch.

**Comment by rhvgoyal** [comment]
I might have pasted 9p, cache=loose numbers wrong previously. This one should be correct.

9p, cache=loose
----------------------

>    READ: bw=39.1MiB/s (41.0MB/s), 39.1MiB/s-39.1MiB/s (41.0MB/s-41.0MB/s), io=
3070MiB (3219MB), run=78471-78471msec
  WRITE: bw=13.1MiB/s (13.7MB/s), 13.1MiB/s-13.1MiB/s (13.7MB/s-13.7MB/s), io=
1026MiB (1076MB), run=78471-78471msec

**Comment by rhvgoyal** [comment]
@ganeshmaharaj are you running virtiofsd with xattr enabled (-o xattr). I removed that and even with cache=always now I am getting around 250MB/s with random writes.

I narrowed it down further and looks like primary slowdown is with cache=always and randwrite only. randreads are just fine. And lot of this slowdown seems to be coming from call file_remove_privs(). It is resulting in lot of getxattr() and that means round extra round trip to server for every write.

**Comment by rhvgoyal** [comment]
I have proposed a patch upstream.

https://lore.kernel.org/linux-fsdevel/20200716144032.GC422759@redhat.com/

**Comment by rhvgoyal** [comment]
Proposed another patch set to fix this issue.

https://lore.kernel.org/linux-fsdevel/20200916161737.38028-1-vgoyal@redhat.com/

**Comment by jcvenegas** [comment]
I have tested qemu virtiofs code buidling from source (no  kata scripts). In the case of kata with 9pfs is more better becasue this patch https://github.com/kata-containers/packaging/blob/stable-1.12/qemu/patches/5.0.x/0001-9p-removing-coroutines-of-9p-to-increase-the-I-O-per.patch

**Comment by jcvenegas** [comment]
@rhvgoyal @dagrh I think virtiofs need some work to have some parity related with that patch.  Seems that is mostly how request are handled.
- I have tested shared-thread pool patch but I did not see difference for this use case.

**Comment by rhvgoyal** [comment]
@jcvenegas I posted V3 of patches to improve random write performance.

https://lore.kernel.org/linux-fsdevel/20201009181512.65496-1-vgoyal@redhat.com/

It has not been merged yet. I hope it will be merge and that should help.

I will look into this qemu out of tree patch you have mentioned. It processes everything in the context of single thread. We think this is not a very scalable approach. IIUC, if this single thread blocks, everything will block behind it. And if we need more threads to introduce parallelism, we can't do that.

Having said that, we probably can introduce something similar for virtiofsd and give it a try and see if it helps.

**Comment by jcvenegas** [comment]
@rhvgoyal thanks - I will try the V3 patches. 
> Having said that, we probably can introduce something similar for virtiofsd and give it a try and see if it helps.
Thanks :) ,  please let me know if in your comparison you are able to see if using that patch makes 9pfs its better than virtiofs. At least for this command:

```
fio --randrepeat=1 --ioengine=libaio --direct=1 --gtod_reduce=1 --name=test --filename=random_read_write.fio --bs=4k --iodepth=64 --size=4G --readwrite=randrw --rwmixread=75
```

**Comment by rhvgoyal** [comment]
@jcvenegas I would say split your tests in two.

- Pure random writes
- Mixed random read and random writes

My patches mentioned above will help a lot with first test. With second test, an additional problem is introduced. I proposed a fix for that upstream but upstream thinks that I need to implement a more generic solution to the problem like file leases.

https://lore.kernel.org/linux-fsdevel/20200929185015.GG220516@redhat.com/

Exploring file leases for fuse is an entirely new feature. I will try to spend some time on that.

**Comment by jcvenegas** [comment]
I ran this test again using https://github.com/rhvgoyal/qemu/commit/823dd12f7a2f393632c86b98e8ee380044282d12

In my worksatation 9pfs still was working better but I have tried in two more servers and it is working better virtiofs.
9pfs:
```
  READ: bw=179MiB/s (187MB/s), 179MiB/s-179MiB/s (187MB/s-187MB/s), io=3070MiB (3219MB), run=17178-17178msec                                                                                                                                                                
  WRITE: bw=59.7MiB/s (62.6MB/s), 59.7MiB/s-59.7MiB/s (62.6MB/s-62.6MB/s), io=1026MiB (1076MB), run=17178-17178msec                            
```
virtiofs:
```
   READ: bw=213MiB/s (223MB/s), 213MiB/s-213MiB/s (223MB/s-223MB/s), io=3070MiB (3219MB), run=14412-14412msec                                                                                                                                                                
  WRITE: bw=71.2MiB/s (74.6MB/s), 71.2MiB/s-71.2MiB/s (74.6MB/s-74.6MB/s), io=1026MiB (1076MB), run=14412-14412msec                                                                                                                                                          
```

I also added runc as part of my comparison to see how well was doing kata based in the container.

Using overlays  I see actually Kata works better than runc.

```
In the case of runc the bandwidth  is more less than kata.
   READ: bw=30.7MiB/s (32.2MB/s), 30.7MiB/s-30.7MiB/s (32.2MB/s-32.2MB/s), io=3070MiB (3219MB), run=99986-99986msec
  WRITE: bw=10.3MiB/s (10.8MB/s), 10.3MiB/s-10.3MiB/s (10.8MB/s-10.8MB/s), io=1026MiB (1076MB), run=99986-99986msec
```

Seems that non of virtiofs or qemu-9pfs are using directIO  (as this tests requires) on the host and probably why is doing faster.

@ Thanks for the fixes on this tests looking forward for the next virtiofs-dev tag in your fork yo upgrade kata.

Cheers

**Comment by rhvgoyal** [comment]
By default virtiofs will not do direct I/O. And I think virtio-9p does not do either. direct I/O specified by test is limited to bypassing guest page cache only (And not host page cache). 

If one needs to avoid host page cache as well, now virtiofsd has introduced a new option to force that. (-o allow_direct_io)

**Comment by jcvenegas** [comment]
awesome, I wonder if we should go by default in Kata. So direct io will be done when the file is open with that flag, correct the rest will do page cache?

**Comment by rhvgoyal** [comment]
If a guest application is doing direct I/O, I think we should avoid page cache on guest. But there is no guarantee that we should avoid page cache on host as well. In storage stack, caching can happen at so many layers. Disks have internal caches, we don't avoid those caches for direct I/O. If a VM is running in cloud, we will not avoid caches on host. So to me, not avoiding caches on host should be default. And if there is a specific use case where it makes sense to avoid cache on host as well, then daemon should be started with correct option.

**Comment by fidencio** [comment]
---

This issue is being automatically closed as Kata Containers 1.x has now reached EOL (End of Life). This means it is **no longer being maintained**.

> **Important:**
>
> All users should switch to the latest [Kata Containers 2.x release](https://github.com/kata-containers/kata-containers/blob/main/docs/install/README.md) to ensure they are using a maintained release that contains the latest security fixes, performance improvements and new features.

This decision was discussed by the @kata-containers/architecture-committee and has been announced via the [Kata Containers mailing list](http://lists.katacontainers.io/pipermail/kata-dev/):
* http://lists.katacontainers.io/pipermail/kata-dev/2020-November/001601.html
* http://lists.katacontainers.io/pipermail/kata-dev/2021-April/001843.html
* http://lists.katacontainers.io/pipermail/kata-dev/2021-May/001896.html

If you believe this issue still applies to Kata Containers 2.x, please open an issue against the [Kata Containers 2.x repository](https://github.com/kata-containers/kata-containers), pointing to this one, providing details to allow us to migrate it.

---
